Update contributing guide with more details on Rust process (#13170)
Qiskit/qiskit: CONTRIBUTING.md

   which will run these checks and report any issues.

   If your code fails the local style checks (specifically the black

>    or rust code formatting check) you can use `tox -eblack` and

Reviewer: I like how you corrected "rust" to "Rust" in other parts of the document, then introduced some more lower case here lol

>    or rust code formatting check) you can use `tox -eblack` and

Reviewer 2: ```suggestion
   or Rust code formatting check) you can use `tox -eblack` and
```

>    `cargo fmt` to automatically fix update the code formatting.

Reviewer 3: ```suggestion
   `cargo fmt` to automatically fix the code formatting.
```

### Testing Rust components

> Many Rust-accelerated functions are generally tested from Python space, but in cases
> where new Rust-native APIs are being added or there is Rust-specific internal details
> to be tested, `#[test]` functions can be included inline.  Typically it's most
> convenient to place these in a separate inline module that is only conditionally
> compiled in, such as

Reviewer 2: ```suggestion
Many Rust-accelerated functions are generally tested from Python space, but in cases
where new Rust-native APIs are being added, or there are Rust-specific internal details
to be tested, `#[test]` functions can be included inline. It's typically most
convenient to place these in a separate inline module that is only conditionally
compiled in, such as
```

}
```

For more detailed guidance on how to add Rust testing you can refer to the Rust
documentation's [guide on writing tests](https://doc.rust-lang.org/book/ch11-01-writing-tests.html)

> for more details and suggestions.

Reviewer 2: ```suggestion
For more detailed guidance on how to add Rust testing you can refer to the Rust
documentation's [guide on writing tests](https://doc.rust-lang.org/book/ch11-01-writing-tests.html).
```

`pip install -r requirements-dev.txt`, then `ruff` and `black` will be available and can be run from
the command line. See [`tox.ini`](tox.ini) for how `tox` invokes them.

### Rust style and lint

> If you're working on Rust code the Rust formatting and lint checking are done
> using different tooling. Qiskit uses [rustfmt](https://github.com/rust-lang/rustfmt) for
> code formatting, and you can simply run `cargo fmt` (if you installed Rust with the
> default settings using `rustup`) and it will update the code formating automatically to
> conform to the style guidelines. This is very similar manner to what `black` and running
> `tox -eblack` for Python code. For lint Qiskit uses [clippy](https://github.com/rust-lang/rust-clippy)
> which can be invoked via `cargo clippy`. For CI to pass you will need clippy to pass
> without any warnings or errors.
> 
> These still get checked via the tox lint commands but if you need to update formating
> you'll need to run ``cargo fmt``.

Reviewer 2: ```suggestion
For formatting and lint checking Rust code, you'll need to use different tools than you would for Python. Qiskit uses [rustfmt](https://github.com/rust-lang/rustfmt) for
code formatting. You can simply run `cargo fmt` (if you installed Rust with the
default settings using `rustup`), and it will update the code formatting automatically to
conform to the style guidelines. This is very similar to running `tox -eblack` for Python code. For lint checking, Qiskit uses [clippy](https://github.com/rust-lang/rust-clippy) which can be invoked via `cargo clippy`. 

Rust lint and formatting checks are included in the the `tox -elint` command. For CI to pass you will need both checks to pass without any warnings or errors. Note that this command checks the code but won't apply any modifications, if you need to update formatting, you'll need to run `cargo fmt`.
```
